Project-Aurora/Project-Aurora/Devices/UnifiedHID/UnifiedHID.cs

```diff
@@ -500,4 +562,460 @@ public bool SetBottomLed(byte r, byte g, byte b)
+
+                hwmap[offset + 0] = src.keys[k].r;
+                hwmap[offset + 12] = src.keys[k].g;
+                hwmap[offset + 24] = src.keys[k].b;
+
+            }
+
+            // First chunk comes with header
+            workbuf[0] = 0x00;
+            workbuf[1] = 0xa1;
+            workbuf[2] = 0x01;
+            workbuf[3] = 0x01;
+            workbuf[4] = 0xb4;
+
+            //memcpy(&workbuf[5], hwmap, 60);
+            Array.Copy(hwmap, 0, workbuf, 5, 60);
+
+            NativeOverlapped overlapped = new NativeOverlapped();
+
+            if (WriteFile(ctrl_device_leds.Handle, workbuf, (uint)workbuf.Length, out _, ref overlapped) != true)
+            {
+                return false;
+            }
+
+            // Six more chunks
+            for (int i = 1; i < 7; i++)
+            {
+                workbuf[0] = 0x00;
+
+                //memcpy(&workbuf[1], &hwmap[(i * 64) - 4], 64);
+                Array.Copy(hwmap, (i * 64) - 4, workbuf, 1, 64);
+
+                if (WriteFile(ctrl_device_leds.Handle, workbuf, (uint)workbuf.Length, out _, ref overlapped) != true)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        // Most of these functions were ported from https://github.com/duncanthrax/roccat-vulcan
+        static bool rv_wait_for_ctrl_device()
+        {
+            for (int i = 1; i < 100; i++) // If still fails after 100 tries then timeout
+            {
+                // 150ms is the magic number here, should suffice on first try.
+                Thread.Sleep(150);
+                bool success = ctrl_device.ReadFeatureData(out byte[] buffer, 0x04);
+                if (success && buffer.Length > 2)
+                {
+                    if (buffer[1] == 0x01)
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        private bool rv_get_ctrl_report(byte report_id)
+        {
+            bool success = ctrl_device.ReadFeatureData(out byte[] buffer, report_id);
+
+            return (success && buffer.Length >= 8);
+        }
+
+        private bool rv_set_ctrl_report(byte report_id)
+        {
+            byte[] buffer = null;
+            //int length = 0;
+
+            switch (report_id)
+            {
+                case 0x15:
+                    buffer = new byte[] { 0x15, 0x00, 0x01 };
+                    //length = 3;
+                    break;
+                case 0x05:
+                    buffer = new byte[] { 0x05, 0x04, 0x00, 0x04 };
+                    //length = 4;
+                    break;
+                case 0x07:
+                    buffer = new byte[] { 0x07 ,0x5f ,0x00 ,0x3a ,0x00 ,0x00 ,0x3b ,0x00 ,0x00 ,0x3c ,0x00 ,0x00 ,0x3d ,0x00 ,0x00 ,0x3e,
```

Comment:
These should all probably be constants rather than being instantiated every time they're needed